python/comps.py

```diff
@@ -0,0 +1,328 @@
+import numpy as np
+import numba as nb
+import pandas as pd
+
+# The number of bins to create for when binning data by sale price; e.g. if
+# this value is 10, the data will be binned into deciles
+NUM_PRICE_BINS = 30
+
+
+def get_comps(
+    observation_df,
+    comparison_df,
+    weights,
+    n=20,
+    num_price_bins=NUM_PRICE_BINS,
+):
+    """Fast algorithm to get the top `n` comps from a dataframe of lightgbm
+    leaf node assignments (`observation_df`) compared to a second dataframe of
+    assignments (`comparison_df`). Leaf nodes are weighted according to a tree
+    importance vector `weights` and used to generate a similarity score and
+    return two dataframes, one a set of indices and the other a set of scores
+    for the `n` most similar comparables. More details on the underlying
+    algorithm here: https://ccao-data.github.io/lightsnip/articles/finding-comps.html
+
+    The function expects that `observation_df` will have a column called
+    `pred_pin_final_fmv` and `comparison_df` will have a column called
+    `meta_sale_price`. These two columns represent the predicted value
+    and the observed value for the observations and the comparisons,
+    respectively. These columns are then used along with the `num_price_bins`
+    parameter to bin the comparison data and only compare observations to
+    comparisons that are in the three closest bins to the observation.
+    """
+    # Convert the weights to a numpy array so that we can take advantage of
+    # numba acceleration later on
+    weights_arr = np.asarray(weights, dtype=np.float64)
+
+    # Add ID columns so that we can keep track of the initial position of
+    # each row as we sort them. This is necessary to allow the caller to
+    # translate results back to parcels based on row index
+    observation_df["id"] = list(range(len(observation_df)))
+    comparison_df["id"] = list(range(len(comparison_df)))
+
+    # Sort the comparison data and extract the indexes of rows that represent
+    # boundaries between price bins. We'll use these bin indexes to
+    # reduce the number of comparison parcels that we need to search for each
+    # observation
+    sorted_comparison_df = comparison_df.sort_values(
+        ["meta_sale_price"]
```

Comment:
**issue:** Swap the sale price binning for predicted value binning, as mentioned above.

Reply:
Done in 06eb0ab9da871cf8f2c9dd9ffc370e78ae19caa9!